heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -2886,13 +2894,18 @@ getDistPerms = use (implStatePerms . distPerms)
 getTopDistPerm :: ExprVar a -> ImplM vars s r (ps :> a) (ps :> a) (ValuePerm a)
 getTopDistPerm x = use (implStatePerms . topDistPerm x)
 
--- | Set the current 'PermSet'
-setPerms :: PermSet ps -> ImplM vars s r ps ps ()
-setPerms perms = implStatePerms .= perms
+-- | Get a sequence of the top @N@ permissions on the stack
+getTopDistPerms :: prx1 ps1 -> RAssign prx2 ps2 ->
+                   ImplM vars s r (ps1 :++: ps2) (ps1 :++: ps2) (DistPerms ps2)
+getTopDistPerms ps1 ps2 = snd <$> RL.split ps1 ps2 <$> getDistPerms
 
--- | Set the current permission for a given variable
-setPerm :: ExprVar a -> ValuePerm a -> ImplM vars s r ps ps ()
-setPerm x p = implStatePerms . varPerm x .= p
+-- | Find all @lowned@ permissions held in in the variable permissions
+implFindLOwnedPerms :: ImplM vars s r ps ps [(ExprVar LifetimeType,
+                                              ValuePerm LifetimeType)]
+implFindLOwnedPerms =
+  mapMaybe (\case NameAndElem l p@(ValPerm_LOwned _ _ _) -> Just (l,p)
```

Comment:
I thought about that, but `ValPerm_LOwned` has two existential variables, and so that change would require either building a special-purpose return type for `implFindLOwnedPerms` or some weird hackery with `Some`, either of which would be a bit gross.